### DISCUSSION TITLE
Typescript: Generic component restricted by value type

### DISCUSSION THREAD

**Discussion by simonedelmann** [react-hook-form/react-hook-form#5489] [Q&A]
Hello! I am currently writing some generic form components. Therefore I would like to restrict a `name` prop to fields with correct type. I hope the example makes clear what I want to achieve:

```ts
// Given a FormData type
type FormData = {
  foo: string
  bar: number
}

// Given a TextField component
type TextFieldProps<FormData> = {
  name: FieldPath<FormData>
}
function TextField<FormData>({name}: TextFieldProps<FormData>) {
  // ...
}

// Using the TextField component
<TextField<FormData> name="foo" /> // is valid
<TextField<FormData> name="bar" /> // is valid, but should be invalid because typeof bar != 'string'
```

Is it possible to restrict `name` to only those paths which conform to the fields type?

I have a Select component in mind, where you can only specify options, which match the given field. Something like below. 

```ts
type SelectFieldProps<FormData> = {
  name: FieldPath<FormData>
  options: Array< /* infer type from FormData and name prop somehow */ >
}
```

I would be glad if anybody could point me into the right direction with this! Thank you very much!

**Comment by bluebill1049** [comment]
I don't think this is possible.
